### PR TITLE
Define OTEL_CONFIG_FILE with placeholder for new environment variable override scheme

### LINKED ISSUE #3752: Should file configuration merge environment variable configuration?
The conversation about whether [file configuration](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md) should completely ignore the [sdk environment variable scheme](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk-environment-variables.md) [came up](https://github.com/open-telemetry/opentelemetry-specification/pull/3744#issuecomment-1785665250) in #3744, but that PR doesn't actually contain any language related to this.

The original [file configuration OTEP](https://github.com/open-telemetry/oteps/blob/main/text/0225-configuration.md#get-tracerprovider-meterprovider-loggerprovider) stated:

> Interpret the [configuration model](https://github.com/open-telemetry/oteps/blob/main/text/0225-configuration.md#configuration-model) and return SDK TracerProvider, MeterProvider, LoggerProvider which strictly reflect the configuration object's details and ignores the [opentelemetry environment variable configuration scheme](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md).

As mentioned [here](https://github.com/open-telemetry/opentelemetry-specification/pull/3744#issuecomment-1786021951), file configuration doesn't actually contain language describing this behavior. It was included originally included in #3437 but was lost in the PR review shuffle - accidentally, not in response to feedback.

@tedsuo argues in favor file configuration respecting env vars with:

> The common expectation among developers is that env vars will automatically overwrite config parameters. If we do it the other way, I am concerned that until the end of time we will have a steady stream of users lodging issues about this and then becoming very frustrated when we explain that they need to modify the config file to use an env var.

> The reason that these users will be upset is that they are in a situation where having to define the env vars in the config file is a non-starter. Their use case is that for some reason they either can't get at or are not allowed to modify the config file template, but they really need to override a parameter.

> For reference, this is an issue that has come up on many OSS projects I have been involved with where it is common to have both operators and application developers wanting to configure the same thing. Often it's some kind of emergency situation where the person with the rights to change the config file is unavailable.

> I should note that of course the opposite situation could be true, where for some reason you want to disable an env var but don't have access to it. But it's probably easier to give users the ability to disable an env var via the config file than it is to give them the ability to disable a config parameter via an env var.

@MrAlias argues in favor of ignoring env vars with:

> To be fair, from experience, you're going to get people complaining either way. If you choose environment variable priority over a configuration users will complain that their deployment was altered and failed when an environment variable was set that took precedence.

> However, if you make environment variables take precedence you will also need to make some pretty sever and subjective choices on how they are mapped to a config. Do the BSP environment variables apply to all batch span processors or just one? Is the sampler environment variable use for all tracer providers in the config, even ones that specify alternate samplers? Should propagators be merged or overridden? If an exporter is defined by environment, does that stop the console logging exporter used for debugging as well as all the other exporters defined in configuration?

> Ultimately, I think the current changes are going to be the most appropriate. They allow users to make their own choice in precedence without making subjective choices for them on how to map things. If a user want environment variables to take precedence, all they need to do is use the OTel environment keys in the related parts of their configuration. In doing so they will answer each of the above questions their own way.

@trask supports the feeling of users expecting env vars to override file configuration, but also says merging configuration from multiple sources is hard:

> This is my feeling as well, especially for things like:

> OTEL_SDK_DISABLED
OTEL_RESOURCE_ATTRIBUTES
OTEL_SERVICE_NAME
OTEL_LOG_LEVEL
OTEL_PROPAGATORS
OTEL_TRACES_SAMPLER_ARG
I totally understand the nightmare that is merging configuration from multiple sources though.

> I wonder if we would have created many of the other env vars (e.g. OTEL_BSP_*, OTEL_BLRP_*) if we had configuration file support from the beginning? And if so, maybe we can deprecate those other env vars in favor of configuration file?

This topic came up several times during the lengthly review of the file configuration OTEP. Below are links to a number of and relevant points:

https://github.com/open-telemetry/oteps/pull/225#discussion_r1116269308

> Layering of config as described below would make it more difficult to reason about what my program is actually being run with.

> Additionally, perhaps give them a helper config file that uses env var substitution in the right places so that they can migrate easily (and still get a warning until they get rid of env variables and move everything to the config file).

https://github.com/open-telemetry/oteps/pull/225#discussion_r1119068865

> What about 'Solution 3: fail when both environment configuration and file configuration are present'? 

> I think we could log a warning when we detect this, but failing is too strong. Consider the implications if a user is operating in an environment they don't fully control (i.e. where an ops team configures environment variables by default which they extend / layer on top of).

https://github.com/open-telemetry/oteps/pull/225#discussion_r1142380977

> I have a (rather strong) opinion that setting set via env vars has higher priority (takes precedence) over a setting set via configiuration file and it should be marked as a goal.

> Any scheme where environment variables have priority over a config file will require some sort of standard mapping between the environment variable schema and file config scheme. IMO, its impossible to define such a mapping which is intuitive in all cases, so better not to try.

> Nothing is forcing users to use file based config - its opt in. If they do opt in, they're opting into the documented behavior in which the config file represents the source of truth for configuration. If they wish to customize the experience with additional layers / overrides, they have a couple of tools:

> 1. They can use the fact the Configure(config) API accepts a config model as an argument, and provide their own customizations to the model after the initial parsing of the file via Parse(file). An example of such a customization would be to interpret environment variables and apply them to the model in a way they decide makes sense.
> 2. They can use environment variable substitution to reference environment variables directly in a configuration file.

---

Update  3/15/2024

The current state of this issue is:
- The configuration working group [proposed and executed a process](https://github.com/open-telemetry/opentelemetry-specification/issues/3752#issuecomment-1972184874) to move forward after several months of failing to reach consensus
- As a part of that process, the configuration maintainers reviewed proposals and have recommended moving forward with option 2.b as described in [this comment](https://github.com/open-telemetry/opentelemetry-specification/issues/3752#issuecomment-1995582317)
- As a part of that process, people who disagree with the recommendation can / should escalate to a TC decision

---

Update 3/28/2024

Please see [this comment](https://github.com/open-telemetry/opentelemetry-specification/pull/3948#issuecomment-2025825640) updating the status of the issue:

> Per @tedsuo’s [request](https://github.com/open-telemetry/opentelemetry-specification/pull/3948#issuecomment-2021052830), we discussed this issue in the 3/24/27 TC meeting and have made a decision:
> Generally, we will follow [@trask's](https://github.com/open-telemetry/opentelemetry-specification/pull/3948#issuecomment-2021332565) comment, proceeding with this PR with a few changes:
> - Rename OTEL_CONFIG_FILE to OTEL_EXPERIMENTAL_CONFIG_FILE, reflecting the fact that the semantics around how the value of env var are subject to breaking changes as the file configuration spec and schema continue to evolve.
> - Ensure that env vars which don’t interop with file config are deprecated when file config is ready for stabilization, reflecting that we do not want to recommend multiple competing configuration stories. This could be ensured via an explicit note in the markdown, or a blocking issue - both achieve the same effect. https://github.com/open-telemetry/opentelemetry-specification/issues/3967
> - Ensure that file config has an interop where platforms (i.e. Azure functions, otel operator, etc) contribute to config. We should proceed with [#3948](https://github.com/open-telemetry/opentelemetry-specification/pull/3948) without being prescriptive about how that mechanism works. In the TC meeting, 4 distinct solutions were discussed which had different tradeoffs and limitations. It is clear that we still need to learn more about the requirements and constraints of this use case and let the findings inform the solution. The config working group should prioritize this discussion, but an answer shouldn’t block this PR. We should open a new issue to track the requirements and discuss solutions, and ensure that we treat that issue as blocking for any sort of stabilization effort (although it should ideally be solved much sooner). https://github.com/open-telemetry/opentelemetry-specification/issues/3966

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #3752.

An alternative to #3805 which ignores the existing environment variable scheme, and leaves a placeholder for defining a new environment variable scheme specific to file configuration, with names derived from the configuration model. 

This is the proposal I suggested [here](https://github.com/open-telemetry/opentelemetry-specification/issues/3752#issuecomment-1915005820). To restate the advantages:

- Avoid the messy merge semantics associated with merging the existing environment variable scheme
- Provide environment variable overrides, which users have come to expect
- We can soften the [sharp edges](https://docs.google.com/document/d/1yPfdf6fsxWY7onWU_PLmIIPs14H6pTYbyI7OboiODCw/edit?pli=1#heading=h.r8b09z6u7mwh) associated with a environment variable scheme derived from the model by refactoring the configuration model to be more friendly to the rules (i.e. by avoiding object arrays)

To understand exactly what a refactored configuration model might look like, I've put together https://github.com/open-telemetry/opentelemetry-configuration/pull/69 to review along side of this.

### CODE REVIEW

File: specification/configuration/sdk-environment-variables.md [hunk 1]
```diff
@@ -288,6 +296,48 @@ that use [periodic exporting MetricReader](../metrics/sdk.md#periodic-exporting-
...
+An example Java implementation:
+
+```java
+OpenTelemetryConfiguration parsedModel = parse("/path/to/config.yaml");
+OpenTelemetryConfiguration modelWithEnvironment = mergeEnvironment(parsedModel);
+if (parsedModel.equals(modelWithEnvironment)) {
+    logger.warn("Environment variables were set that overrode the configuration specified at /path/to/config.yaml");
+}
+// Log at debug level so user has to opt-in to logging effective configuration
+logger.debug("Effective configuration: \n" + toString(modelWithEnvironment));
```
Comment (brunobat) [inline specification/configuration/sdk-environment-variables.md:331]: Please beware that printing config values can expose passwords and secrets. If there is no way to opt-out specific config values, the values should not be printed at all.

Comment (jack-berg) [reply specification/configuration/sdk-environment-variables.md:331]: In this example the user is opting out by default, since we use a verbose / non-default log level. The user opts in by configuring their log level to be more verbose. If they happen to use the verbose log level by default, they can opt increasing the threshold log level or turning off logs altogether from the relevant logger. The comment above this line explains this.

Comment (brunobat) [reply specification/configuration/sdk-environment-variables.md:331]: ok, thanks

File: specification/configuration/sdk-environment-variables.md [hunk 2]
```diff
@@ -288,6 +296,48 @@ that use [periodic exporting MetricReader](../metrics/sdk.md#periodic-exporting-
...
+
+| Name             | Description                                                                                                                        | Default | Notes     |
+|------------------|------------------------------------------------------------------------------------------------------------------------------------|---------|-----------|
+| OTEL_CONFIG_FILE | The path of the configuration file used to configure the SDK. If set other environment variables defined in this file are ignored. |         | See below |
+
+If `OTEL_CONFIG_FILE` is set, the file at the specified path is used to
+call [Parse](file-configuration.md#parse), [Merge Environment](./file-configuration#merge-environment), [Create](file-configuration.md#create).
+The implementation SHOULD log a warning if calling Merge Environment results in
+changes to
+the [configuration model](./file-configuration.md#configuration-model).
```
Comment (trask) [inline specification/configuration/sdk-environment-variables.md:313]: I didn't follow why emit a warning for this?

**Review by trask — APPROVED** [review]
I still have questions around the existing env vars, but I support this PR as a good incremental step forward 👍

File: specification/configuration/sdk-environment-variables.md (hunk 2 above)
Comment (jack-berg) [reply specification/configuration/sdk-environment-variables.md]: I think its important to draw attention to the fact that the contents of the file don't reflect the configuration of the SDK. I imagine the workflow being something like:
- Include an overriding environment variable
- Receive a warning that the configuration file is different (warning doesn't mean a problem in this context - more like a notification): "Environment variables were set that overrode the configuration specified at /path/to/config.yaml"
- If opted in, log the entire effective configuration after applying overrides so you can deduce what changed

Comment (trask) [reply specification/configuration/sdk-environment-variables.md]: I'd suggest limiting this to a debug message. It could be worth looking at Spring for prior art.

Comment (jack-berg) [reply specification/configuration/sdk-environment-variables.md]: > I'd suggest limiting this to a debug message

That's fine. I can relax the language to say "The implementation SHOULD emit a log if calling Merge Environment results in change...". Leave it to implementations to decide the appropriate log level.

**Review by MrAlias — APPROVED** [review]

**Review by marcalff — CHANGES_REQUESTED** [review]
I would like to make a radical suggestion.

Yes, the configuration is expressed in a config.yaml file, to be consumed by each runtime for each SIG, no change.

Yes, the configuration can be merged with additional environment variables if we really want to. This is a separate discussion, but no change as far as this PR is concerned.

However, now that it is established that this merge operation is independent of the existing environment variables implemented by each SIG SDK, the merge operation becomes SIG agnostic.

Hence, I don't see why the merge should be implemented in each language, in each sdk, and embedded in each runtime.

The proposal is to perform the merge operation in a separate process, and only feed the merged configuration yaml to the opentelemetry sdk.

This as the benefit of:

* only implement merge once, not for each language SIG
* implement the merge as a separate process, and not bloat every runtime
* provide more stability to runtimes

In short, the output of a merge is a final config.yaml, and no merge operation is to occur in memory with the configuration model inside each SDK.

This has another far reaching consequence:

* for extension points, the merge is done up front, in yaml
* the implementation of each extension point (custom exporter, sampler, etc) will not have to implement a merge operation on the custom model either, making things much simpler for people providing extension points.

**Comment by tedsuo** [comment]
@marcalff unfortunately running a separate process is a non-starter for many end user environments that OpenTelemetry needs to support.

**Review by tedsuo — APPROVED** [review]
I agree that this is the correct approach for moving forwards. Thanks for threading the needle on this one.

**Review by marcalff — COMMENTED** [review]
Previous suggestion withdrawn - it does not work for all SIG.

The is unfortunate, as a lot of complexity will be pushed to every SIG runtime.

**Review by marcalff — APPROVED** [review]

File: specification/configuration/sdk-environment-variables.md [hunk 3]
```diff
@@ -288,6 +296,48 @@ that use [periodic exporting MetricReader](../metrics/sdk.md#periodic-exporting-
...
 
+## File Configuration
+
+**Status**: [Experimental](../document-status.md)
+
+Environment variables involved in [file configuration](./file-configuration.md).
+
+| Name             | Description                                                                                                                        | Default | Notes     |
+|------------------|------------------------------------------------------------------------------------------------------------------------------------|---------|-----------|
+| OTEL_CONFIG_FILE | The path of the configuration file used to configure the SDK. If set other environment variables defined in this file are ignored. |         | See below |
```
Comment (brunobat) [inline specification/configuration/sdk-environment-variables.md:307]: >  If set other environment variables defined in this file are ignored.

It needs to be clarified in which file will the environment variables be ignored. The file defined in OTEL_CONFIG_FILE?

**Review by brunobat — CHANGES_REQUESTED** [review]
